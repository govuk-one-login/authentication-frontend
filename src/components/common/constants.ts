import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
