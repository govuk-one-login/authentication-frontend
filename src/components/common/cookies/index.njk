{% extends "common/layout/base.njk" %}
{% from "govuk/components/table/macro.njk" import govukTable %}
{% from "govuk/components/radios/macro.njk" import govukRadios %}
{% from "govuk/components/button/macro.njk" import govukButton %}
{% from "govuk/components/notification-banner/macro.njk" import govukNotificationBanner %}

{% set pageTitleName = 'pages.cookiePolicy.title' | translate %}

{% block content %}

{% set html %}
<h3 class="govuk-notification-banner__heading">
  {{ 'pages.cookiePolicy.successBanner.header' | translate }}
</h3>
<p class="govuk-body">{{ 'pages.cookiePolicy.successBanner.paragraph1' | translate }}</p>
<a class="govuk-notification-banner__link" id="go-back-link" href="#">{{ 'pages.cookiePolicy.successBanner.linkText' | translate }}</a>
{% endset %}

{% if updated %}
{{
  govukNotificationBanner({
  html: html,
  type: 'success',
  titleText: 'general.cookie.cookieBanner.cookieBannerTitleText' | translate,
  attributes: {
    "id": "save-success-banner"
  }
}) }}
{% else %}
{{ govukNotificationBanner({
  html: html,
  type: 'success',
  titleText: 'general.cookie.cookieBanner.cookieBannerTitleText' | translate,
  attributes: {
    "id": "save-success-banner",
    "hidden": true
  }
}) }}
{% endif %}

<h1 class="govuk-heading-l govuk-!-margin-top-0 govuk-!-margin-bottom-3">
  {{ 'pages.cookiePolicy.header' | translate }}
</h1>

<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph1' | translate}}
  <a href="{{'pages.cookiePolicy.info.govUkCookiesLinkHref' | translate}}" class="govuk-link" rel="noreferrer noopener" target="_blank">{{'pages.cookiePolicy.info.govUkCookiesLinkText' | translate}}</a>{{'pages.cookiePolicy.info.paragraph2' | translate}}
</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph3' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph4' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph5' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph6' | translate}}
  <a href="{{'pages.cookiePolicy.info.linkHref' | translate}}" class="govuk-link" rel="noreferrer noopener" target="_blank">{{'pages.cookiePolicy.info.linkText' | translate}}</a>{{'pages.cookiePolicy.info.paragraph7' | translate}}
</p>

<h2 class="govuk-heading-s">{{'pages.cookiePolicy.essential.header' | translate}}</h2>

<p class="govuk-body">{{'pages.cookiePolicy.essential.info.paragraph1' | translate}}</p>
<ul class="govuk-list govuk-list--bullet">
  <li>{{'pages.cookiePolicy.essential.info.bulletPoint1' | translate}}</li>
  <li>{{'pages.cookiePolicy.essential.info.bulletPoint2' | translate}}</li>
  <li>{{'pages.cookiePolicy.essential.info.bulletPoint3' | translate}}</li>
</ul>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
    text: 'pages.cookiePolicy.essential.info.table.headers.col1' | translate,
    classes: 'govuk-body-s'
    },
    {
    text: 'pages.cookiePolicy.essential.info.table.headers.col2' | translate,
    classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
      {
      text: "aps",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "_csrf",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row2.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "gs",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row3.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "di-persistent-session-id",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row4.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "am",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row5.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "lo",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row6.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "hmpo-wizard-sc",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row7.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "service_session",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row8.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "cri_passport_service_session",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row9.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "ipv_core_service_session",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row10.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "chl",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row11.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "re",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row12.col2' | translate,
      classes: 'govuk-body-s'
      }
    ]
  ]
}) }}

<h3 class="govuk-heading-s">{{'pages.cookiePolicy.cookiesMessage.header' | translate}}</h3>
<p class="govuk-body">{{'pages.cookiePolicy.cookiesMessage.paragraph1' | translate}}</p>
<ul class="govuk-list govuk-list--bullet">
  <li>{{'pages.cookiePolicy.cookiesMessage.bulletPoint1' | translate}}</li>
  <li>{{'pages.cookiePolicy.cookiesMessage.bulletPoint2' | translate}}</li>
</ul>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.headers.col1' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.headers.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.headers.col3' | translate,
      classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
    {
      text: "cookies_preferences_set",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.rows.row1.col3' | translate,
      classes: 'govuk-body-s'
    }
    ]
  ]
}) }}

<h2 class="govuk-heading-s">{{'pages.cookiePolicy.settingsCookies.header' | translate}}</h2>
<p class="govuk-body">{{'pages.cookiePolicy.settingsCookies.paragraph1' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.settingsCookies.paragraph2' | translate}}</p>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
      text: 'pages.cookiePolicy.settingsCookies.table.headers.col1' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.headers.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.headers.col3' | translate,
      classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
    {
      text: "lng",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.rows.row1.col3' | translate,
      classes: 'govuk-body-s'
    }
    ]
  ]
}) }}

<h2 class="govuk-heading-s">{{'pages.cookiePolicy.analytics.header' | translate}}</h2>
<p class="govuk-body">{{'pages.cookiePolicy.analytics.info.paragraph1' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.analytics.info.paragraph2' | translate}}</p>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
      text: 'pages.cookiePolicy.analytics.table.headers.col1' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.headers.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.headers.col3' | translate,
      classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
    {
      text: "_ga",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row1.col3' | translate,
      classes: 'govuk-body-s'
    }
    ],
    [
    {
      text: "_gid",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row2.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row2.col3' | translate,
      classes: 'govuk-body-s'
    }
    ],
    [
    {
      text: "_gat_UA-[number]",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row3.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row3.col3' | translate,
      classes: 'govuk-body-s'
    }
    ]
  ]
}) }}

<form method="post" id="cookie-preferences-form" novalidate hidden="true">
<input type="hidden" name="_csrf" value="{{csrfToken}}"/>
{{ govukRadios({
  name: "cookie_preferences",
  attributes: {
    "id": "radio-cookie-preferences"
  },
  items: [
    {
      value: "true",
      text: 'pages.cookiePolicy.settings.yesRadioButton' | translate,
      id: "policy-cookies-accepted",
      checked: analyticsConsent === true
    },
    {
      value: "false",
      text: 'pages.cookiePolicy.settings.noRadioButton' | translate,
      id: "policy-cookies-rejected",
      checked: analyticsConsent === false
    }
  ]
}) }}

{{ govukButton({
  "text": 'pages.cookiePolicy.settings.saveButton' | translate,
  "type": "submit",
  "preventDoubleClick": true,
  attributes: {
    "id": "save-cookie-settings"
  }
}) }}

</form>

{% endblock %}
