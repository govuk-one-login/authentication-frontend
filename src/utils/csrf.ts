import { csrfSync } from "csrf-sync";

const { csrfSynchronisedProtection } = csrfSync({
  getTokenFromRequest: (req) => {
    return req.body["_csrf"];
  },
});

export { csrfSynchronisedProtection };
