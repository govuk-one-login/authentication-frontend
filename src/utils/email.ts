export function redactEmail(value: string): string {
  const asterisks = "***";
  let prefix = "";
  let domain = "";

  const email_fragments = value.split("@");

  if (email_fragments.length === 2) {
    [prefix, domain] = email_fragments;
    return `${prefix.substring(0, 3)}${asterisks}@${domain}`;
  }
  return "";
}
