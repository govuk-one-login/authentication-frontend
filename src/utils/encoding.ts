import { InvalidBase64Error } from "./error.js";
export function base64DecodeToUint8Array(value: string): Uint8Array {
  const padding = "=".repeat((4 - (value.length % 4)) % 4);
  const base64 = (value + padding).replace(/-/g, "+").replace(/_/g, "/");

  const isBase64 =
    /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)?$/.test(base64);
  if (!isBase64) {
    throw new InvalidBase64Error("String is not valid base64: " + value);
  }

  const rawData = Buffer.from(base64, "base64");
  return new Uint8Array(rawData);
}
