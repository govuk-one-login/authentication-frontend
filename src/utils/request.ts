import type { Request } from "express";
import { supportAccountRecovery, supportReauthentication } from "../config.js";
import {
  CONTACT_US_THEMES,
  SERVICE_TYPE,
  SUPPORT_TYPE,
} from "../app.constants.js";
export const isReauth = (req: Request): boolean =>
  supportReauthentication() && Boolean(req.session?.user?.reauthenticate);

export const isUpliftRequired = (req: Request): boolean =>
  Boolean(req.session?.user?.isUpliftRequired);

export const isAccountRecoveryJourney = (req: Request): boolean =>
  Boolean(req.session?.user?.isAccountRecoveryJourney);

export const isAccountRecoveryJourneyAndEnabled = (req: Request): boolean =>
  Boolean(
    supportAccountRecovery() &&
      req.session?.user?.isAccountRecoveryJourney &&
      req.session?.user?.isAccountRecoveryPermitted
  );

export const isContactUsSuggestionsFeedbackSubtheme = (req: Request): boolean =>
  req.query?.subtheme === CONTACT_US_THEMES.SUGGESTIONS_FEEDBACK;

export const clientIsOneLogin = (req: Request): boolean =>
  Boolean(req.session?.client?.isOneLoginService);

export const clientUsesOneLoginOptionally = (req: Request): boolean =>
  req.session?.client?.serviceType === SERVICE_TYPE.OPTIONAL;

export const supportTypeIsGovService = (req: Request): boolean =>
  Boolean(
    req.query?.supportType === SUPPORT_TYPE.GOV_SERVICE ||
      req.body?.supportType === SUPPORT_TYPE.GOV_SERVICE
  );

export const urlContains = (req: Request, str: string): boolean =>
  Boolean(req.originalUrl?.includes(str));
