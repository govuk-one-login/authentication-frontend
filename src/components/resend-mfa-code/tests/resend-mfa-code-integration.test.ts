import request from "supertest";
import { describe } from "mocha";
import { sinon } from "../../../../test/utils/test-utils";
import nock = require("nock");
import * as cheerio from "cheerio";
import decache from "decache";
import {
  API_ENDPOINTS,
  HTTP_STATUS_CODES,
  PATH_NAMES,
} from "../../../app.constants";
import { ERROR_CODES } from "../../common/constants";

describe("Integration:: resend mfa code", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = require("../../../middleware/session-middleware");
    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (req: any, res: any, next: any): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";

        req.session.user = {
          email: "test@test.com",
          phoneNumber: "7867",
          journey: {
            nextPath: PATH_NAMES.ENTER_MFA,
            optionalPaths: [PATH_NAMES.RESEND_MFA_CODE],
          },
        };

        next();
      });

    app = await require("../../../app").createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;

    await request(app)
      .get(PATH_NAMES.RESEND_MFA_CODE)
      .then((res) => {
        const $ = cheerio.load(res.text);
        token = $("[name=_csrf]").val();
        cookies = res.headers["set-cookie"];
      });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  it("should return resend mfa code page", (done) => {
    request(app).get(PATH_NAMES.RESEND_MFA_CODE).expect(200, done);
  });

  it("should return error when csrf not present", (done) => {
    request(app)
      .post(PATH_NAMES.RESEND_MFA_CODE)
      .type("form")
      .send({
        code: "123456",
      })
      .expect(500, done);
  });

  it("should redirect to /enter-code when new code requested as part of sign in journey", (done) => {
    nock(baseApi)
      .post(API_ENDPOINTS.MFA)
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT);

    request(app)
      .post(PATH_NAMES.RESEND_MFA_CODE)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
      })
      .expect("Location", PATH_NAMES.ENTER_MFA)
      .expect(302, done);
  });

  it("should redirect to /check-your-phone when new code requested as part of account creation journey", (done) => {
    nock(baseApi)
      .post(API_ENDPOINTS.MFA)
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT);

    request(app)
      .post(PATH_NAMES.RESEND_MFA_CODE)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        isResendCodeRequest: true,
      })
      .expect("Location", PATH_NAMES.CHECK_YOUR_PHONE)
      .expect(302, done);
  });

  it("should render 'You cannot get a new security code at the moment' when OTP lockout timer cookie is active", () => {
    const testSpecificCookies = cookies + "; re=true";
    request(app)
      .get(PATH_NAMES.RESEND_MFA_CODE)
      .set("Cookie", testSpecificCookies)
      .expect((res) => {
        res.text.includes("You cannot get a new security code at the moment");
      });
  });

  it("should return 500 error screen when API call fails", (done) => {
    nock(baseApi).post(API_ENDPOINTS.MFA).once().reply(500, {
      errorCode: "1234",
    });

    request(app)
      .post(PATH_NAMES.RESEND_MFA_CODE)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
      })
      .expect(500, done);
  });

  it("should redirect to /security-code-requested-too-many-times when request OTP more than 5 times", (done) => {
    nock(baseApi)
      .post(API_ENDPOINTS.MFA)
      .times(6)
      .reply(400, { code: ERROR_CODES.MFA_SMS_MAX_CODES_SENT });

    request(app)
      .post(PATH_NAMES.RESEND_MFA_CODE)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
      })
      .expect(
        "Location",
        "/security-code-requested-too-many-times?actionType=mfaMaxCodesSent"
      )
      .expect(302, done);
  });

  it("should redirect to /security-code-invalid-request when exceeded OTP request limit", (done) => {
    nock(baseApi)
      .post(API_ENDPOINTS.MFA)
      .once()
      .reply(400, { code: ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED });

    request(app)
      .post(PATH_NAMES.RESEND_MFA_CODE)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
      })
      .expect(
        "Location",
        "/security-code-invalid-request?actionType=mfaBlocked"
      )
      .expect(302, done);
  });
});
