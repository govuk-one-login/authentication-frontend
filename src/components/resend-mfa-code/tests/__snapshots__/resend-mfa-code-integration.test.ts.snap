// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration:: resend mfa code should include the last three digits of the user's telephone number 1`] = `
Object {
  "contentId": "f463a280-31f1-43c0-a2f5-6b46b1e2bb15",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should render 'You cannot get a new security code at the moment' when OTP lockout timer cookie is active 1`] = `
Object {
  "contentId": "f463a280-31f1-43c0-a2f5-6b46b1e2bb15",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should return 400 error screen when API call fails 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should return 500 error screen when API call fails 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should return error when csrf not present 1`] = `
Object {
  "contentId": undefined,
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
  "taxonomyLevel3": undefined,
  "taxonomyLevel4": undefined,
  "taxonomyLevel5": undefined,
}
`;

exports[`Integration:: resend mfa code should return resend mfa code page with reauth analytics properties 1`] = `
Object {
  "contentId": "a2776ef7-9ef3-4d8d-bdbc-3f798b15e5d4",
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "re auth",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should return resend mfa code page with sign in analytics properties 1`] = `
Object {
  "contentId": "f463a280-31f1-43c0-a2f5-6b46b1e2bb15",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should state reauthenticating user could be logged out 1`] = `
Object {
  "contentId": "a2776ef7-9ef3-4d8d-bdbc-3f798b15e5d4",
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "re auth",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration:: resend mfa code should state user could be locked out 1`] = `
Object {
  "contentId": "f463a280-31f1-43c0-a2f5-6b46b1e2bb15",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
